urllib3/sync_connection.py

```diff
@@ -0,0 +1,556 @@
+                conn, ssl_context, fingerprint, assert_hostname
+            )
+
+        # Now that the connection is created, we want to set the socket to
+        # non-blocking mode. We're going to select on it for the rest of its
+        # lifetime, so we need it non-blocking. We also register it with our
+        # selector to allow us to assume that it is *always* registered.
+        conn.setblocking(0)
+        self._sock = conn
+        self._selector.register(
+            self._sock, selectors.EVENT_READ | selectors.EVENT_WRITE
+        )
+
+    def send_request(self, request, read_timeout):
+        """
+        Sends a single Request object. Returns a Response.
+        """
+        # TODO: Replace read_timeout with something smarter.
+        self._read_timeout = read_timeout
+
+        # Before we begin, confirm that the state machine is ok.
+        assert self._state_machine.our_state is h11.IDLE
+        assert self._state_machine.their_state is h11.IDLE
+
+        # First, register the socket with the selector. We want to look for
+        # readability *and* writability, because if the socket suddenly becomes
+        # readable we need to stop our upload immediately.
+        self._selector.modify(
+            self._sock, selectors.EVENT_READ | selectors.EVENT_WRITE
+        )
+        header_bytes = _request_to_bytes(request, self._state_machine)
+        body_chunks = _body_bytes(request, self._state_machine)
+        request_chunks = itertools.chain([header_bytes], body_chunks)
+
+        for chunk in request_chunks:
+            # If the socket becomes readable we don't need to error out or
+            # anything: we can just continue with our current logic.
+            readable = self._send_unless_readable(chunk)
+            if readable:
+                break
+
+        # At this point we no longer care if the socket is writable.
+        self._selector.modify(self._sock, selectors.EVENT_READ)
+
+        response = None
+        while response is None:
+            read_bytes = self._receive_bytes(read_timeout)
+            response = _maybe_read_response(read_bytes, self._state_machine)
+
+        if response.http_version not in _SUPPORTED_VERSIONS:
+            raise BadVersionError(response.http_version)
+
+        version = b'HTTP/' + response.http_version
+        our_response = Response(
+            status_code=response.status_code,
+            headers=_headers_to_native_string(response.headers),
+            body=self,
+            version=version
+        )
+        return our_response
+
+    def close(self):
+        """
+        Close this connection, suitable for being re-added to a connection
+        pool.
+        """
+        if self._sock is not None:
+            sock, self._sock = self._sock, None
+            sock.setblocking(1)
+            sock.close()
+
+        if self._selector is not None:
+            selector, self._selector = self._selector, None
```

Comment:
It's actually more than that, it's to ensure that the selector is removed from the object so that it cannot subsequently be accessed. It essentially guarantees that once we've closed a connection there is a defined state.